File: 2026/08-August/05-Wednesday/solution.py
from typing import List

class Solution:
    def remainingMethods(self, n: int, k: int, invocations: List[List[int]]) -> List[int]:
        graph = [[] for _ in range(n)]
        undirected = [[] for _ in range(n)]

        for u, v in invocations:
            graph[u].append(v)
            undirected[u].append(v)
            undirected[v].append(u)

        suspicious = [False] * n

        def dfs(node):
            suspicious[node] = True
            for nxt in graph[node]:
                if not suspicious[nxt]:
                    dfs(nxt)

        dfs(k)

        visited = [False] * n

        def dfs2(node):
            visited[node] = True
            for nxt in undirected[node]:
                if not visited[nxt]:
                    suspicious[nxt] = False
                    dfs2(nxt)

        for i in range(n):
            if not suspicious[i] and not visited[i]:
                dfs2(i)

        return [i for i in range(n) if not suspicious[i]]

File: 2026/08-August/05-Wednesday/solution.java
import java.util.*;

class Solution {
    public List<Integer> remainingMethods(int n, int k, int[][] invocations) {
        List<Integer>[] graph = new ArrayList[n];
        List<Integer>[] undirected = new ArrayList[n];

        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
            undirected[i] = new ArrayList<>();
        }

        for (int[] e : invocations) {
            int u = e[0];
            int v = e[1];
            graph[u].add(v);
            undirected[u].add(v);
            undirected[v].add(u);
        }

        boolean[] suspicious = new boolean[n];
        dfs(k, graph, suspicious);

        boolean[] visited = new boolean[n];

        for (int i = 0; i < n; i++) {
            if (!suspicious[i] && !visited[i]) {
                dfs2(i, undirected, suspicious, visited);
            }
        }

        List<Integer> ans = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            if (!suspicious[i]) {
                ans.add(i);
            }
        }

        return ans;
    }

    private void dfs(int node, List<Integer>[] graph, boolean[] suspicious) {
        suspicious[node] = true;

        for (int nxt : graph[node]) {
            if (!suspicious[nxt]) {
                dfs(nxt, graph, suspicious);
            }
        }
    }

    private void dfs2(int node, List<Integer>[] graph,
                      boolean[] suspicious, boolean[] visited) {

        visited[node] = true;

        for (int nxt : graph[node]) {
            if (!visited[nxt]) {
                suspicious[nxt] = false;
                dfs2(nxt, graph, suspicious, visited);
            }
        }
    }
}
